src/darwin/Framework/CHIP/MTRDiagnosticLogsDownloader.mm

```diff
@@ -0,0 +1,630 @@
+    VerifyOrReturn(nil != download, completion([MTRError errorForCHIPErrorCode:CHIP_ERROR_NOT_FOUND]));
+
+    download.abortHandler = abortHandler;
+    completion(nil);
+}
+
+- (void)handleBDXTransferSessionDataForFileDesignator:(NSString *)fileDesignator
+                                          fabricIndex:(NSNumber *)fabricIndex
+                                               nodeID:(NSNumber *)nodeID
+                                                 data:(NSData *)data
+                                           completion:(MTRStatusCompletion)completion
+{
+    assertChipStackLockedByCurrentThread();
+    MTR_LOG_DEFAULT("BDX Transfer Session Data: %@: %@", fileDesignator, data);
+
+    auto * download = [_downloads get:fileDesignator fabricIndex:fabricIndex nodeID:nodeID];
+    VerifyOrReturn(nil != download, completion([MTRError errorForCHIPErrorCode:CHIP_ERROR_NOT_FOUND]));
+
+    NSError * error = nil;
+    [download writeToFile:data error:&error];
+    VerifyOrReturn(nil != error, completion(error));
+
+    completion(nil);
+}
+
+- (void)handleBDXTransferSessionEndForFileDesignator:(NSString *)fileDesignator
+                                         fabricIndex:(NSNumber *)fabricIndex
+                                              nodeID:(NSNumber *)nodeID
+                                               error:(NSError * _Nullable)error
+{
+    assertChipStackLockedByCurrentThread();
+    MTR_LOG_DEFAULT("BDX Transfer Session End: %@: %@", fileDesignator, error);
+
+    auto * download = [_downloads get:fileDesignator fabricIndex:fabricIndex nodeID:nodeID];
+    VerifyOrReturn(nil != download);
+
+    VerifyOrReturn(nil == error, [download failure:error]);
+    [download success];
+}
+@end
+
+DiagnosticLogsDownloaderBridge::DiagnosticLogsDownloaderBridge(MTRDiagnosticLogsDownloader * delegate)
+{
+    mDelegate = delegate;
+}
+
+DiagnosticLogsDownloaderBridge::~DiagnosticLogsDownloaderBridge()
+{
+    mDelegate = nil;
+}
+
+CHIP_ERROR DiagnosticLogsDownloaderBridge::OnTransferBegin(chip::bdx::BDXTransferProxy * transfer)
+{
+    VerifyOrReturnError(nil != mDelegate, CHIP_ERROR_INCORRECT_STATE);
+
+    auto fileDesignatorSpan = transfer->GetFileDesignator();
+    auto fileDesignator = AsString(fileDesignatorSpan);
+    VerifyOrReturnError(nil != fileDesignator, CHIP_ERROR_INCORRECT_STATE);
+
+    auto * fabricIndex = @(transfer->GetFabricIndex());
+    auto * nodeId = @(transfer->GetPeerNodeId());
+
+    auto completionHandler = ^(NSError * _Nullable error) {
+        assertChipStackLockedByCurrentThread();
+
+        if (error != nil) {
+            auto err = [MTRError errorToCHIPErrorCode:error];
+            transfer->Reject(err);
+        } else {
+            transfer->Accept();
+        }
+    };
+
+    // Ideally we would like to handle aborts a bit differently since this only works
+    // because our BDX stack supports one transfer at a time.
```

Comment:
Um... Why does it do that?  It shouldn't do that.  Nivi's PR started with things like that, but was getting into shape where it wasn't doing that....

This is not workable in practice; if Home asks for logs from two devices, or two logs from the same device, what exactly happens?